File: common/src/main/java/org/sandboxpowered/silica/network/serverbound/LoginStart.java
package org.sandboxpowered.silica.network.serverbound;

import org.sandboxpowered.silica.network.Packet;
import org.sandboxpowered.silica.network.PacketByteBuf;
import org.sandboxpowered.silica.network.PacketHandler;
import org.sandboxpowered.silica.network.clientbound.EncryptionRequest;
import org.sandboxpowered.silica.server.SilicaServer;

public class LoginStart implements Packet {
    private String username;

    @Override
    public void read(PacketByteBuf buf) {
        username = buf.readString(16);
    }

    @Override
    public void write(PacketByteBuf buf) {
        buf.writeString(username);
    }

    @Override
    public void handle(PacketHandler packetHandler, SilicaServer server) {
        packetHandler.sendPacket(new EncryptionRequest("", server.getKeyPair().getPublic().getEncoded(), server.getVerificationArray()));
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/vanilla/network/login/serverbound/LoginStart.kt
package org.sandboxpowered.silica.vanilla.network.login.serverbound

import org.sandboxpowered.silica.vanilla.network.Connection
import org.sandboxpowered.silica.vanilla.network.Packet
import org.sandboxpowered.silica.vanilla.network.PacketByteBuf
import org.sandboxpowered.silica.vanilla.network.PacketHandler

class LoginStart(private var username: String) : Packet {
    constructor(buf: PacketByteBuf) : this(buf.readString(16))

    override fun write(buf: PacketByteBuf) {
        buf.writeString(username)
    }

    override fun handle(packetHandler: PacketHandler, connection: Connection) {
        connection.handleLoginStart(username)
    }
}